---
title: |
    | Alignment of Genome Sequencing Data
subtitle: "GSND 5340Q, BMDA"
author: | 
  | W. Evan Johnson, Ph.D.
  | Professor, Division of Infectious Disease
  | Director, Center for Data Science
  | Rutgers University -- New Jersey Medical School
date: "`r Sys.Date()`"
header-includes:
  - \usepackage{amsmath}
  - \usepackage{xcolor}
  - \setbeamercolor{frametitle}{fg=black}
  - \usepackage{graphicx}
  - \usebackgroundtemplate{\includegraphics[width=\paperwidth]{mapFigs/RH_template_Page_2.png}}
  - \addtobeamertemplate{frametitle}{\vspace*{.25in}}{\vspace*{.25in}}
  - \setbeamerfont{frametitle}{size=\huge}
  - \usepackage{tikz}
output: 
  beamer_presentation
classoption: aspectratio=169 
editor_options: 
  chunk_output_type: console
tables: true
---


## Sequencing data alignment
\Large
Next-Generation Sequencing (NGS) Data present new challenges:

- Map `reads' to genome
- Call SNPs and variants from the reads
- Other Applications: RNA-seq, miRNAs, alternative splicing, ChIP-seq, BS-seq, RNA editing

## Sequencing data alignment
\Large
Obstacles:

- Massive data size
- Repeat regions, rare variants
	
**Goal:** Develop an approach to put the puzzle together! 

## Mapping Algorithms
\large
Alignment algorithms (not comprehensive -- some of my favorites!)

- BLAST
- BLAT
- SOAP2
- Bowtie2
- BWA
- STAR
- Subread (Rsubread)
- GNUMAP
- NovoAlign
- Many others!!

## Basic Local Alignment Search Tool (BLAST)
\Large
BLAST is a "heuristic" method that identifies and combined _High-scoring Segment Pairs (HSPs)_ between two sequences. BLAST searches for high scoring sequence alignments between the query sequence and the existing sequences in the database using an approach that approximates the Smith-Waterman algorithm (described later). 

## Basic Local Alignment Search Tool (BLAST)
\center 
![](mapFigs/Neighbor_HSP.jpg){width=70%}

## Basic Local Alignment Search Tool (BLAST)
\Large
BLAST can be found at: [https://blast.ncbi.nlm.nih.gov/Blast.cgiBLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi). Try "BLASTING" the following sequences: 

\scriptsize
`>NC_045512.2 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1`
`ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAA`

`>NC_001802.1 Human immunodeficiency virus 1`
`GGTCTCTCTGGTTAGACCAGATCTGAGCCTGGGAGCTCTCTGGCTAACTAGGGAACCCACTGCTTAAGCCTCAATAAAGC`

`>NR_102810.2 Mycobacterium tuberculosis H37Rv 16S ribosomal RNA`
`TGTTTGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGA`
`AAGGTCTCTTCGGAGATACTCGAGTGGCGAACGGGTGAGTAACACGTGGGTGATCTGCCCTGCACTTCGG`

`>NR_041248.1 Bacillus anthracis strain ATCC 14578 16S ribosomal RNA`
`AGAGCTTGCTCTTATGAAGTTAGCGGCGGACGGGTGAGTAACACGTGGGTAACCTGCCCATAAGACTGGG`

## Creating a Hash Index Table
\center
![](mapFigs/hash1.png){height=70%}

## Creating a Hash Index Table
\center
![](mapFigs/hash2.png){height=70%}

## Creating a Hash Index Table
\center
![](mapFigs/hash3.png){height=70%}

## Creating a Hash Index Table
\center
![](mapFigs/hash4.png){height=70%}

## Creating a Hash Index Table
\center
![](mapFigs/hash5.png){height=70%}

## Read Lookup
\center
![](mapFigs/read1.png){height=70%}

## Read Lookup
\center
![](mapFigs/read2.png){height=70%}

## Read Lookup
\center
![](mapFigs/read3.png){height=70%}

## Read Lookup
\center
![](mapFigs/read4.png){height=70%}

## Read Lookup
\center
![](mapFigs/read5.png){height=70%}

## Needleman-Wunsch
\center
![](mapFigs/needleman_wunsch1.png){width=60%}

## Needleman-Wunsch
\center
![](mapFigs/needleman_wunsch2.png){width=60%}

## Needleman-Wunsch
\center
![](mapFigs/needleman_wunsch3.png){width=60%}

## Needleman-Wunsch
\center
![](mapFigs/needleman_wunsch4.png){width=60%}

## Prefix Trees
\center
![](mapFigs/trie.png){height=80%}


## Prefix Trees
\center
![](mapFigs/bwatrie.png){height=70%}


## Burrows-Wheeler Transformation (BWA, Bowtie)
\center
![](mapFigs/bwa.jpg){height=70%}

## Burrows-Wheeler Transformation (BWA, Bowtie)
\center
![](mapFigs/bowtie.pdf){height=70%}


## Hybrid Methods (BWA-SW, Bowtie2)
\center
![](mapFigs/bowtie2flow.jpg){height=70%}

## Read alignments are stored in SAM/BAM/CRAM

\Large
- BAM and CRAM are compressed SAM files
- One line per alignment
- Use `samtools` to view, sort, merge, concatenate, index, get statistics, etc, on alignment files.

## SAM/BAM/CRAM Format -- Header

\Large
The _header_ lines start with @:

- @HD = header definition
- @SQ = a sequence in the reference file you used, followed by how long it was and itâ€™s comment (from the reference file)
- @RG = read groups you assigned while mapping the reads
- @PG = programs used to obtain this bam, in order


## SAM/BAM/CRAM Format -- Body
\Large
Each alignment has 11 mandatory fields:

\center
![](mapFigs/samformat.png)

## SAM/BAM/CRAM Format -- Flag
![](mapFigs/sam_flag.png)

## SAM/BAM/CRAM Format -- Example
![](mapFigs/sam_example.png)

## SAM/BAM/CRAM Format 
\large

- Use `samtools` view to see the content of a SAM/BAM/CRAM file.
- Always sort and index them
- Use `samtools view -f` (include) and `-F` (exclude) to filter by flags
- Use `samtools view -q` to filter by quality. 
- More on `samtools` later!!

## BWA Example (OnDemand Amarel Desktop)

\large
```{bash, eval=F}
# load bwa
module load bwa

# Index the genome:
bwa index genomefile.fa 

# Align the reads:
bwa mem genomefile.fa myfastqfile.fq > bwaalign.sam

# Note: for use with gatk (will learn later) bwa mem 
# should include a `read group' header as well, e.g.: 
-R "@RG\tID:group1\tSM:sam1\tPL:illumina\tLB:lib1\tPU:unit1" 
```


## Bowtie2 Example (OnDemand Amarel Desktop)

\large
```{bash, eval=F}
# Load Bowtie2
module load bowtie2

# Index the genome:
bowtie2-build genomefile.fa output/genomefile	

# Align the reads:
bowtie2 -x output/genomefile -U myfastqfile.fq \
	-S bowtiealign.sam
```

## RSubread Example (OnDemand RStudio)
\large
```{r, eval=F}
# Install the Bioconductor installer
install.packages("BiocManager")

# Install Rsubread 
BiocManager::install("Rsubread")

# Index the genome:
buildindex("genomefile", "genomefile.fa")

# Align the reads:
align("genomefile", "myfastqfile.fq", 
      output_file="myfastq.bam", nthreads = 4)

```


## Session info
\tiny
```{r session info}
sessionInfo()
```

